crates/re_viewer/src/ui/selection_panel.rs

```diff
@@ -413,6 +408,226 @@ fn entity_props_ui(
+fn visible_history_ui(
+    ctx: &mut ViewerContext<'_>,
+    ui: &mut Ui,
+    space_view_class_name: &SpaceViewClassName,
+    entity_path: Option<&EntityPath>,
+    entity_props: &mut EntityProperties,
+) {
+    //TODO(#4107): support more space view types.
+    if space_view_class_name != "3D" && space_view_class_name != "2D" {
+        return;
+    }
+
+    if !should_display_visible_history(ctx, entity_path) {
+        return;
+    }
+
+    let re_ui = ctx.re_ui;
+
+    re_ui.checkbox(
+        ui,
+        &mut entity_props.visible_history.enabled,
+        "Visible history",
+    ).on_hover_text(
+        "Enable the Visible History feature.\n\nBy default, only the last data received before the \
+        current time is shown. By activating the Visible History, all the data received within a \
+        time window is shown instead.");
+
+    let time_range = if let Some(times) = ctx
+        .store_db
+        .time_histogram(ctx.rec_cfg.time_ctrl.timeline())
+    {
+        times.min_key().unwrap_or_default()..=times.max_key().unwrap_or_default()
+    } else {
+        0..=0
+    };
+
+    let current_time = ctx
+        .rec_cfg
+        .time_ctrl
+        .time_i64()
+        .unwrap_or_default()
+        .at_least(*time_range.start()); // accounts for timeless time (TimeInt::BEGINNING)
+
+    let sequence_timeline = matches!(ctx.rec_cfg.time_ctrl.timeline().typ(), TimeType::Sequence);
+
+    let visible_history = if sequence_timeline {
+        &mut entity_props.visible_history.sequences
+    } else {
+        &mut entity_props.visible_history.nanos
+    };
+
+    ui.add_enabled_ui(entity_props.visible_history.enabled, |ui| {
+        egui::Grid::new("visible_history_boundaries")
+            .num_columns(4)
+            .show(ui, |ui| {
+                ui.label("From");
+                visible_history_boundary_ui(
+                    re_ui,
+                    ui,
+                    &mut visible_history.from,
+                    sequence_timeline,
+                    current_time,
+                    time_range.clone(),
+                    true,
+                );
+
+                ui.end_row();
+
+                ui.label("To");
+                visible_history_boundary_ui(
+                    re_ui,
+                    ui,
+                    &mut visible_history.to,
+                    sequence_timeline,
+                    current_time,
+                    time_range,
+                    false,
+                );
+
+                ui.end_row();
+            });
+    });
+    ui.add(
+        egui::Label::new(
+            egui::RichText::new(if sequence_timeline {
+                "These settings apply to all sequence timelines."
+            } else {
+                "These settings apply to all temporal timelines."
+            })
+            .italics()
+            .weak(),
+        )
+        .wrap(true),
+    );
+}
+
+static VISIBLE_HISTORY_COMPONENT_NAMES: once_cell::sync::Lazy<Vec<ComponentName>> =
```

Comment:
might be better to opt-out instead?
overall it should be clear that this is temporary. If we think it isn't (that there's always things where there's no visible history) we should find a way to traitify this.

Reply:
I dont think so, because you could have e.g. a Transform _and_ a Point3D at the same entity path, so it must be opt-in. (Again, unless I'm missing something.)

Reply:
oh good point. Hmm messy 🤔

Reply:
https://github.com/rerun-io/rerun/issues/4145

Reply:
I also added a TODO there.